### ISSUE TITLE
wall clock time is hours slower in more recent versions

### ISSUE THREAD

**Issue by danstoner** [dimitri/pgcopydb#870]
While transferring a database larger than 10 TB, I was able to complete the copy in 16h with `v0.14`.

With newer versions (`v0.16` and above?), the same process takes 1d7h.

Here is the command I am using:

```
pgcopydb clone --restart --skip-vacuum --split-tables-larger-than 10 GB --table-jobs 48 --index-jobs 32 ...
```

(I arrived at these numbers through a lot of testing to get the total time as short as possible.  The local DB instance and the target instances are quite large to handle this load).

Any ideas why the newer versions would be this much slower?

I am going to test again with `v0.15` because I kinda suspect the new Boehm-Demers-Weiser Garbage Collector that was introduced in `v0.16` is the source of the performance difference.

Alternatively, I am worried that the copy made with `v0.14` is missing some of my data!

**Comment by danstoner** [comment]
My testing of `v0.15` failed with:

```
free(): corrupted unsorted chunks
free(): corrupted unsorted chunks
free(): corrupted unsorted chunks
...
free(): corrupted unsorted chunks
14:48:17.624 54 ERROR  Sub-process 57 exited with code 0 and signal Aborted
14:48:17.624 54 ERROR  Sub-process 58 exited with code 0 and signal Aborted
14:48:17.624 54 ERROR  Sub-process 59 exited with code 0 and signal Aborted
...
14:48:17.625 54 ERROR  Some COPY worker process(es) have exited with error, see above for details
14:48:17.625 54 ERROR  Failed to copy table data, see above for details
14:48:17.627 16 ERROR  Sub-process 54 exited with code 12
14:48:17.928 16 ERROR  Some sub-processes have exited with error status, see above for details
14:48:17.928 16 ERROR  Errors detected, see above for details
14:48:17.928 16 FATAL  Failed to COPY the data, see above for details
14:48:17.928 16 ERROR  Failed to clone source database, see above for details
14:48:18.066 1 ERROR  clone process 16 has terminated [6]
```

Command:
```
docker run --network=host --rm -it dimitri/pgcopydb:v0.15 pgcopydb  clone --restart --skip-vacuum --split-tables-larger-than "10 GB"   --table-jobs 48 --index-jobs 32 ...
```

So I can't say whether the performance difference comes in `v0.15` or `v0.16`+.

**Comment by alexalex89** [comment]
+1 here!

We also made the experience that cloning with < v0.15 (?) seems way faster. Our database is about 40 GB and v0.17 takes so long to clone it that we cancelled it after about 3 hours.
We first thought that it might be caused by non-optimized db-settings and we're also moving from Zalando Postgres Operator to CloudNative PG in the same step.
After finding @danstoner issue we tried downgrading which made things much faster.

One can spot a difference in growing database size.
v0.17:
![image](https://github.com/user-attachments/assets/10cb2bb0-6858-454f-86f7-2bd5c281bf75)

v0.14.1:
![image](https://github.com/user-attachments/assets/1f80518d-59ca-4dbd-a6dc-683a7c156d22)

**Comment by dimitri** [comment]
This might have something to do with the new ability to expose transmitted bytes while COPY commands are running. I can't think of another change that would have an impact on the TABLE DATA copy phase in v0.17. Is it possible to run some `perf` profiling to see more about where the slowness might come from?

**Comment by mmoscher** [comment]
@dimitri yes this should be possible. I'm a colleague of @alexalex89. Could you support us with a suitable command?

Furthermore, we assume that it's not the amount of data (TABLE DATA?) which cause this issue, but the amount of tables or schemas.

Some days ago we successfully copied a database using v0.17 with following stats:

```
2024-08-20 15:01:44.745 8 INFO   main.c:136                Running pgcopydb version 0.17-1.pgdg120+1 from "/usr/bin/pgcopydb"

....

2024-08-20 15:01:44.859 10 INFO   copydb_schema.c:761       Fetched information for 184 tables (including 0 tables split in 0 partitions total), with an estimated total of 5041 thousands tuples and 7165 MB on-disk
2024-08-20 15:01:44.875 10 INFO   copydb_schema.c:968       Fetched information for 258 indexes (supporting 198 constraints)
2024-08-20 15:01:44.881 10 INFO   sequences.c:78            Fetching information for 53 sequences
2024-08-20 15:01:44.898 10 INFO   copydb_schema.c:1122      Fetched information for 5 extensions
2024-08-20 15:01:44.964 10 INFO   copydb_schema.c:1538      Found 0 indexes (supporting 0 constraints) in the target database
````

this took us arround 3mins.
Running the same command against larger table with way more tables and indexes wasn't able to complete in 8 hours (this was our longest try).

```
2024-08-29 14:55:21.273 10 INFO   copydb_schema.c:761       Fetched information for 211152 tables (including 0 tables split in 0 partitions total), with an estimated total of 58 million tuples and 47 GB on-disk
2024-08-29 14:55:34.662 10 INFO   copydb_schema.c:968       Fetched information for 317651 indexes (supporting 239171 constraints)
2024-08-29 14:55:37.531 10 INFO   sequences.c:78            Fetching information for 35598 sequences
2024-08-29 14:57:35.209 10 INFO   copydb_schema.c:1122      Fetched information for 5 extensions
2024-08-29 14:58:11.270 10 INFO   copydb_schema.c:1538      Found 0 indexes (supporting 0 constraints) in the target database
```

However, if you could support us with a suitable command, we can try to run some profiling tomorrow.

**Comment by mmoscher** [comment]
Hi @dimitri,

anything we could assist with? I look forward to hearing from you.

Cheer's

**Comment by dimitri** [comment]
We traced the regression in PR #999. The performance problem sits entirely in the SQLite catalog layer that was introduced in v0.16 — not in the GC or the COPY path itself, as confirmed by the 211K-table / 317K-index numbers @mmoscher shared above.

Four changes are in the fix:

### 1. Two new SQLite indexes that fix the O(N²) scan

`catalog_s_table_count_indexes()` is called once per table during the copy phase to decide whether a table's indexes have all been created. With 211K tables and 317K rows in the `s_index` table, and no index on the lookup column, each call did a full sequential scan — 211K × 317K = ~67 billion row comparisons in the worst case.

Two new indexes reduce each lookup from O(N) to O(log N):

```sql
CREATE INDEX s_i_tableoid    ON s_index(tableoid)
CREATE INDEX s_c_indexoid    ON s_constraint(indexoid)
```

### 2. Deferred index creation

Previously, SQLite indexes were created on empty tables at startup and then maintained row-by-row during the bulk catalog load. Building an index on a fully-populated table is 3–5× faster than incremental maintenance. The two new indexes above are now created *after* the bulk schema fetch completes via a new `catalog_create_indexes()` call.

### 3. Prepared statement cache for hot-path reads

SQLite compiles a SQL string into a bytecode object with `sqlite3_prepare_v2()` and frees it with `sqlite3_finalize()`. `catalog_s_table_count_indexes()` — called 211K times — was compiling and freeing the same statement on every call. A small `uthash`-keyed cache (`CachedStmt`) now keeps compiled statements alive and reuses them with `sqlite3_reset()` + `sqlite3_clear_bindings()`, eliminating 211K compile/deallocate cycles.

### 4. Multi-row batch INSERTs for the catalog write path

The schema fetch inserts one row at a time into `s_table`, `s_attr`, `s_index`, and `s_constraint`. With 317K indexes and 211K tables (each with several attributes), that's millions of individual INSERT round-trips to SQLite. The fix accumulates rows in a heap buffer and flushes them as a single multi-row `VALUES (…), (…), …` statement, bounded by SQLite's parameter limit (32766 on SQLite 3.40):

| Table          | Columns | Max batch size |
|----------------|---------|----------------|
| s_table        | 11      | 2978 rows      |
| s_attr         | 10      | 3276 rows      |
| s_index        | 10      | 3276 rows      |
| s_constraint   |  6      | 5461 rows      |

---

The net effect for the 211K-table database: the catalog phase that previously took hours now completes in seconds, and the wall-clock time for the full clone is back in line with v0.14.

**Comment by runhardr** [comment]
Amazing work!  Thank you!  ❤️
